### PR TITLE
Add a new flag --platform to acr build and task

### PR DESCRIPTION
**Pull request by shahzzzam**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by shahzzzam** [comment]
@djyou @northtyphoon Can you guys review this?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 1]
```diff
@@ -90,6 +90,7 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('no_wait', help="Do not wait for the run to complete and return immediately after queuing the run.", action='store_true')
         c.argument('no_format', help="Indicates whether the logs should be displayed in raw format", action='store_true')
         c.argument('os_type', options_list=['--os'], help='The operating system type required for the build.', arg_type=get_enum_type(OsType))
+        c.argument('platform', options_list=['--platform'], help='The platform required for this build.')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Options_list not needed here since the CLI will automatically infer this.

File: src/command_modules/azure-cli-acr/HISTORY.rst [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 Release History
 ===============
+2.1.14
+++++++
+* Add support for Platform flag.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/HISTORY.rst]: Be more specific. Example:
```
* `acr build`: Add `--platform` argument. 
```

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 3]
```diff
@@ -107,12 +106,14 @@ def acr_build(cmd,
             is_push_enabled = False
             logger.warning("'--image or -t' is not provided. Skipping image push after build.")
 
+    platform_os, platform_arch = get_validate_platform(os_type, platform)
+    logger.warning("OS is %s and Architecture is %s", platform_os, platform_arch)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Why is this a warning?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 1 above)
Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: doone

File: src/command_modules/azure-cli-acr/HISTORY.rst (hunk 2 above)
Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/HISTORY.rst]: done

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 3 above)
Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: fixed

**Comment by shahzzzam** [comment]
> Please provide an example of the usage here. Something seems off. Also, a flag takes no value. The are new arguments, not new flags.

Hi @tjprescott , Could you be more specific? What is off about this?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 4]
```diff
@@ -327,10 +330,17 @@ def acr_task_update(cmd,  # pylint: disable=too-many-locals
                 name=base_image_trigger.name if base_image_trigger else "defaultBaseimageTriggerName"
             )
 
+    platform_os = None
+    platform_arch = None
+    if os_type or platform:
```
Comment (northtyphoon) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Do you need to check here? Can `get_validate_platform` handle the `None` case and provide the default value?

File: src/command_modules/azure-cli-acr/HISTORY.rst [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 Release History
 ===============
+2.1.14
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/HISTORY.rst]: Is `2.1.13` released?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 6]
```diff
@@ -90,6 +90,7 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('no_wait', help="Do not wait for the run to complete and return immediately after queuing the run.", action='store_true')
         c.argument('no_format', help="Indicates whether the logs should be displayed in raw format", action='store_true')
         c.argument('os_type', options_list=['--os'], help='The operating system type required for the build.', arg_type=get_enum_type(OsType))
+        c.argument('platform', help='The platform required for this build.')
```
[the remark this answers is not in the document]
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: It might be better to provide examples in `_help.py`.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py [hunk 7]
```diff
@@ -269,6 +269,48 @@ def validate_managed_registry(cli_ctx, registry_name, resource_group_name=None,
     return registry, resource_group_name
 
 
+def get_validate_platform(os_type, platform):
+    """Gets and validates the Platform from both flags
+    :param str os_type: The name of OS passed by user in --os flag
+    :param str platform: The name of Platform passed by user in --platform flag
+    """
+    from azure.mgmt.containerregistry.v2018_09_01.models import OS, Architecture
```
[the remark this answers is not in the document]
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: Let's keep this as is for now. There is an item to do a cleanup https://github.com/Azure/azure-cli/issues/8265

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py [hunk 8]
```diff
@@ -269,6 +269,48 @@ def validate_managed_registry(cli_ctx, registry_name, resource_group_name=None,
...
+        platform_split = platform.split('/', 1)
+        platform_os = platform_split[0]
+        platform_arch = platform_split[1] if len(platform_split) > 1 else Architecture.amd64.value
+
+    if os_type and platform:
+        if os_type.lower() != platform_os.lower():
+            raise CLIError("The OS in '--platform' should exactly match the value provided in '--os'")
+    elif os_type:
+        platform_os = os_type
+    else:
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: NIT: not needed since you are passing.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 9]
```diff
@@ -44,7 +42,8 @@ def acr_build(cmd,
               no_format=False,
               no_push=False,
               no_logs=False,
-              os_type=OS.linux.value):
+              os_type='',
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Default to `None`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 10]
```diff
@@ -107,12 +106,14 @@ def acr_build(cmd,
             is_push_enabled = False
             logger.warning("'--image or -t' is not provided. Skipping image push after build.")
 
+    platform_os, platform_arch = get_validate_platform(os_type, platform)
+    logger.info("OS is %s and Architecture is %s", platform_os, platform_arch)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: How is this log used?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjprescott) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: I agree. Please add at least one example of this usage to the _help.py file.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py (hunk 4 above)
Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: yes it can

File: src/command_modules/azure-cli-acr/HISTORY.rst (hunk 5 above)
Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/HISTORY.rst]: How to check that?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: done

**Comment by shahzzzam** [comment]
@tjprescott is the CI broken on Unittest step for everyone?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 11]
```diff
@@ -437,15 +437,15 @@
...
         - name: Create a Linux task from a public GitHub repository which builds the hello-world image with a git commit trigger
           text: >
             az acr task create -t hello-world:{{.Run.ID}} -n hello-world -r MyRegistry -c https://github.com/Azure-Samples/acr-build-helloworld-node.git -f Dockerfile --git-access-token 0000000000000000000000000000000000000000
-        - name: Create a Windows task from a public GitHub repository which builds the Azure Container Builder image.
+        - name: Create a Windows task from a public GitHub repository which builds the Azure Container Builder image on Arm/v7 architecture.
           text: >
-            az acr task create -t acb:{{.Run.ID}} -n acb-win -r MyRegistry -c https://github.com/Azure/acr-builder.git -f Windows.Dockerfile --commit-trigger-enabled false --pull-request-trigger-enabled false --os Windows
+            az acr task create -t acb:{{.Run.ID}} -n acb-win -r MyRegistry -c https://github.com/Azure/acr-builder.git -f Windows.Dockerfile --commit-trigger-enabled false --pull-request-trigger-enabled false --platform Windows/arm/v7
```
Comment (northtyphoon) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: We cannot build windows for arm arch. Can you change it to a valid case like linux/arm/v7?

File: src/command_modules/azure-cli-acr/HISTORY.rst [hunk 12]
```diff
@@ -2,10 +2,10 @@
 
 Release History
 ===============
-
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/HISTORY.rst]: Unnecessary change

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py [hunk 13]
```diff
@@ -269,6 +269,54 @@ def validate_managed_registry(cli_ctx, registry_name, resource_group_name=None,
...
+
+    if platform:
+        platform_split = platform.split('/')
+        platform_os = platform_split[0]
+        platform_arch = platform_split[1] if len(platform_split) > 1 else Architecture.amd64.value
+        platform_variant = platform_split[2] if len(platform_split) > 2 else None
+
+    if os_type and platform:
+        if os_type.lower() != platform_os.lower():
+            raise CLIError("The OS in '--platform' should exactly match the value provided in '--os'")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: NIT: add an ending period

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py [hunk 14]
```diff
@@ -269,6 +269,54 @@ def validate_managed_registry(cli_ctx, registry_name, resource_group_name=None,
...
+    elif os_type:
+        platform_os = os_type
+
+    platform_os = platform_os.title()
+    platform_arch = platform_arch.title()
+    valid_os = [item.value.title() for item in OS]
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: Create const for these?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 15]
```diff
@@ -327,10 +330,18 @@ def acr_task_update(cmd,  # pylint: disable=too-many-locals
...
+    platform_variant = None
+    if os_type or platform:
+        platform_os, platform_arch, platform_variant = get_validate_platform(os_type, platform)
+
     taskUpdateParameters = TaskUpdateParameters(
         status=status,
         platform=PlatformUpdateParameters(
-            os=os_type
+            os=platform_os if platform_os else os_type,
+            architecture=platform_arch if platform_arch else None,
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Can just be `architecture=platform_arch`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 16]
```diff
@@ -327,10 +330,18 @@ def acr_task_update(cmd,  # pylint: disable=too-many-locals
...
+            variant=platform_variant if platform_variant else None
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Can just be `variant=platform_variant`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py [hunk 17]
```diff
@@ -269,6 +272,51 @@ def validate_managed_registry(cli_ctx, registry_name, resource_group_name=None,
...
+
+    if platform:
+        platform_split = platform.split('/')
+        platform_os = platform_split[0]
+        platform_arch = platform_split[1] if len(platform_split) > 1 else Architecture.amd64.value
+        platform_variant = platform_split[2] if len(platform_split) > 2 else None
+
+    if os_type and platform:
+        if os_type.lower() != platform_os.lower():
+            raise CLIError("The OS in '--platform' should exactly match the value provided in '--os'.")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: This was what I was looking for in the example. If platform contains the OS then why the redundancy?  It would seem you could just accept, `--os` and `--platform`, each with reflected enums, and then concatenate them together if that's the format the service needs. Otherwise, how is someone supposed to know what values of platform they can use?

Comment (shahzzzam) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: I think that's because we wanted to be consistent with docker's own `--platform` flag (which is currently experimental) https://docs.docker.com/engine/reference/commandline/build/#options
Docker does not have any `os` flag though. I am not sure if the plan is to deprecate `--os` from az acr. //cc: @northtyphoon

Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: Yes. We try to provide similar user experience as `docker build` command in `az acr build` command. That's why we introduce a new `--platform` build parameter and it's very specific to build scenarios. But we still need to keep `--os` for backward compatible. It's possible to deprecate `--os` in the future. I think the code paves the way for the future transition.

Comment (tjprescott) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: In that case, you can deprecate `--os` now. Just mark it as hidden. It will continue to work but point people toward the new parameter. Also, it should help prevent scenarios where the user supplies both.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 18]
```diff
@@ -89,8 +89,8 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('no_logs', help="Do not show logs after successfully queuing the build.", action='store_true')
         c.argument('no_wait', help="Do not wait for the run to complete and return immediately after queuing the run.", action='store_true')
         c.argument('no_format', help="Indicates whether the logs should be displayed in raw format", action='store_true')
-        c.argument('os_type', options_list=['--os'], help='The operating system type required for the build.', arg_type=get_enum_type(OsType))
-        c.argument('platform', help='The platform of the resulting image.')
+        c.argument('os_type', options_list=['--os'], help='The operating system type required for the build.', arg_type=get_enum_type(OsType), deprecate_info=c.deprecate(hide=True))
+        c.argument('platform', help="The platform of the resulting image in the format of 'os/arch/variant'. This replaces the '--os' flag which is soon to be deprecated. E.g. linux/arm/v7. The 'arch' and 'variant' parts are optional.")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: For deprecate_info, add a redirect to platform. Don't describe the replacement in your help text.  https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#deprecating-commands-and-arguments

**Comment by northtyphoon** [comment]
@tjprescott can you please merge the PR? 

/cc: @djyou @ehotinger
